core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -436,17 +436,22 @@ abstract class AbstractControllerBrokerRequestBatch(config: KafkaConfig,
   def addStopReplicaRequestForBrokers(brokerIds: Seq[Int],
                                       topicPartition: TopicPartition,
                                       deletePartition: Boolean): Unit = {
-    // A sentinel (-2) is used as an epoch if the topic is queued for deletion. It overrides
-    // any existing epoch.
-    val leaderEpoch = metadataInstance.leaderEpoch(topicPartition)
-
     brokerIds.filter(_ >= 0).foreach { brokerId =>
       val result = stopReplicaRequestMap.getOrElseUpdate(brokerId, mutable.Map.empty)
-      val alreadyDelete = result.get(topicPartition).exists(_.deletePartition)
+      val updatedDeletePartition = deletePartition || result.get(topicPartition).exists(_.deletePartition)
+
+      // A sentinel (-2) is used as an epoch if the replica is to be deleted.
+      // It overrides any existing epoch.
+      val leaderEpoch = if (updatedDeletePartition) {
```

Comment:
Just to clarify this change is now including `result.get(topicPartition).exists(_.deletePartition)` as a condition to mark the epoch as -2? The rest is just cleaning up/simplifying.

Reply:
Basically what I wanted is to have -2 sent as the leader epoch whenever the delete flag was enabled in the `StopReplica` request. That seemed simpler than trying to pinpoint specific cases, such as reassignment cancellation.

Reply:
I guess my question is whether it covers cases of reassignment cancellation when the topic is not deleted. I wasn't sure if delete partition is true in that case.

Reply:
Yeah, `DeletePartition` is set whenever the replica should be deleted, which would be the case after cancellation for all adding replicas. It does not necessarily imply topic deletion.